### DISCUSSION TITLE
Add a transformer options in the persist middleware

### DISCUSSION THREAD

**Discussion by 4msar** [pmndrs/zustand#2247] [Ideas]
The idea behind this was an issue I have faced recently.
Let me explain why this is required.

I have a store with multiple slices (followed slice pattern) now I'm going to persist the store in my browser local store. so I used the persist middleware in my store.

In my slices, one of the slice is containing a Date instance, when I persist the store it was converted as a date string.
Now I where I have use this I have to convert every-time from the date string to Date instance I can do it by the selector like this.
```js
const fromDate = useMyStore(state => new Date(state.fromDate))
```
but i have to to make this each time I use.

for this reason I have created another middleware to do that after I did not find the transformer options in the persist middleware options, there was a persist option which is allow what data should persist the middleware but there was no transformer.

**Comment by dai-shi** [comment]
For such use cases, we suggest using `storage` option. You can create your own storage. This also applies to adding validation.

**Reply by 4msar** [reply]
Yes storage option can be used, but I don't want to touch the default storage, instead of a transformer.

**Reply by dbritto-dev** [reply]
@4msar that's why we have the storage option and that's the option you can use to achieve whatever you want

**Reply by dai-shi** [reply]
The default storage is not type safe, so it's better to have a custom storage for production. I think we need to clarify that in docs. I also wish someone create a third-party library to cover more use cases.

btw, `reviver` and `replacer` options might be helpful for serializing `Date`. Docs should also be updated.

**Reply by 4msar** [reply]
there were no proper documentation about `reviver` or `replacer` and in my current implementation, we have multiple store based on id, and the data is not specified, so it's better to add a transform functionality so we can use individual store based on the data.
and yes, currently we are using another custom middleware to transform the data, if the reviver or replacer can do the same then we may use it.

**Reply by dai-shi** [reply]
Apologies for no docs. We probably forgot to add it when we added the capability.

@dbritto-dev @charkour Would you guys help update the docs?
@4msar You can also help the project directly too.

**Reply by 4msar** [reply]
@dai-shi yaah, I will IA.

**Reply by charkour** [reply]
@dai-shi, I'm struggling to understand why `retriever` and `replacer` exist when we have the custom `storage` option where the zustand user can effectively apply `retriever` and `replacer` on a [per-field level within the persisted store](https://github.com/pmndrs/zustand/blob/main/docs/integrations/persisting-store-data.md#how-do-i-use-it-with-map-and-set). 

I saw a conversation on the PR here where you initially state `storage` was supposed to be sufficient for this use case: https://github.com/pmndrs/zustand/pull/1763

**Reply by dai-shi** [reply]
In a sense, I agree. But, we already have `createJSONStorage` for convenience (which is technically unnecessary), and it's convenient to have `reviver` and `replacer` there.

**Reply by charkour** [reply]
I've added a PR here to update docs to include `reviver` and `replacer`: https://github.com/pmndrs/zustand/pull/2299
